skimage/feature/texture.py

```diff
@@ -97,11 +99,21 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
     assert_nD(distances, 1, 'distances')
     assert_nD(angles, 1, 'angles')
 
-    assert levels <= 256
     image = np.ascontiguousarray(image)
     assert image.min() >= 0
-    assert image.max() < levels
-    image = image.astype(np.uint8)
+
+    image_max = image.max()
```

Comment:
> You are right in principle, but in practice there are many cases where the full range is not used (i.e. most levels are used in [0, max], but max << max(datatype)).

We have parameter `levels` for this, don't we?

> 2e32 x 2e32

Yes, sorry, the total size of a full GLCM for uin16 will be, at least, twice bigger, because we have uint32 for counters, so `>= (2**16)*(2**16) * 4 / 1024**3 = 16 (GB)`.

---

OK, I can see two ways to go, at the moment:

1. We allow the function to work on full uint16 range __only__ if parameter `labels` is specified. If it is not, we return an error with detailed reasoning (or warning, actually);
2. We add parameter `return_sparse` (or something like this) to the function, implement sparse GLCM computation and allow uint16 images to be used only with `return_sparse=True`.

In both alternatives, the default `labels` value is 2**img.dtype, i.e. full range.

We can get some inspiration from MATLAB (http://www.mathworks.com/help/images/ref/graycomatrix.html). They explicitly clip (`GrayLimits`) and bin (`NumLevels`). Defaults are reasonable (`If I is an intensity image, graycomatrix scales the image to eight gray-levels.`) as they declare even uint64 images support.